### ISSUE TITLE
Return predictions as json

### ISSUE THREAD

**Issue by robmarkcole** [ultralytics/yolov5#2703]
## 🚀 Feature
For creating API around models it would be very useful to add a `to_json()` method for inference results on `class Detections`.

## Motivation
Creating a flask app to expose inferencing, will need to implement a `to_json()` method but preferable to wrap around a standardised result.

## Pitch
Ideally something like:
```
detections = model(img, size=640)
return jsonify({"detections": detections.to_json()})
```

## Additional context
Can be used in https://github.com/ultralytics/yolov5/issues/2662

**Comment by harry418** [comment]
It will be great if yolov5 return a csv or json file of predicted results.

**Comment by glenn-jocher** [comment]
@robmarkcole right, this is a good idea. Do you know what the best JSON format we should use is though? Is there a standardized or common JSON format typically used for this?

The results themselves exist in a few different formats currently, though the default label format (xywh normalized) is the one that we use for saving with the .save() method. 

Also what about passing additional metadata like class names etc in the JSON?
https://github.com/ultralytics/yolov5/blob/9ccfa85249a2409d311bdf2e817f99377e135091/models/common.py#L295-L309

**Comment by glenn-jocher** [comment]
@harry418 you can save inference results to space-delimited text files with the results.save() method:

```python
import torch

# Model
model = torch.hub.load('ultralytics/yolov5', 'yolov5s')

# Images
dir = 'https://github.com/ultralytics/yolov5/raw/master/data/images/'
imgs = [dir + f for f in ('zidane.jpg', 'bus.jpg')]  # batch of images

# Inference
results = model(imgs)
results.save()  # or .print(), .show()
```

**Comment by robmarkcole** [comment]
@glenn-jocher I am not aware of any standards for the kind of JSON response. You could look at the format returned by deepstack, which is using yolo5 already https://docs.deepstack.cc/object-detection/index.html

**Comment by glenn-jocher** [comment]
@robmarkcole sure, that sounds reasonable. Their format is here:
```
{'predictions': [{'x_max': 819, 'x_min': 633, 'y_min': 354, 'confidence': 99, 'label': 'dog', 'y_max': 546}, {'x_max': 601, 'x_min': 440, 'y_min': 116, 'confidence': 99, 'label': 'person', 'y_max': 516}, {'x_max': 445, 'x_min': 295, 'y_min': 84, 'confidence': 99, 'label': 'person', 'y_max': 514}], 'success': True}
```

**Comment by robmarkcole** [comment]
Correct. Conveniently I think the predictions list could be dumped straight to a pandas dataframe too

**Comment by harry418** [comment]
> @harry418 you can save inference results to space-delimited text files with the results.save() method:
> 
> ```python
> import torch
> 
> # Model
> model = torch.hub.load('ultralytics/yolov5', 'yolov5s')
> 
> # Images
> dir = 'https://github.com/ultralytics/yolov5/raw/master/data/images/'
> imgs = [dir + f for f in ('zidane.jpg', 'bus.jpg')]  # batch of images
> 
> # Inference
> results = model(imgs)
> results.save()  # or .print(), .show()
> ```

Thanks @glenn-jocher . i was trying using
 !python detection.py -- source test-file
lol

**Comment by glenn-jocher** [comment]
@harry418 ah, yes you can do it from detect.py also. For saving txt results from detect.py:
```
python detect.py --save-txt
```

**Comment by glenn-jocher** [comment]
@robmarkcole I added pandas support in PR #2725, i.e.

```python
print(results.pandas().xyxy[0])

         xmin        ymin        xmax        ymax  confidence  class    name
0   57.068970  391.770599  241.383545  905.797852    0.868964      0  person
1  667.661255  399.303589  810.000000  881.396667    0.851888      0  person
2  222.878387  414.774231  343.804474  857.825073    0.838376      0  person
3    4.205386  234.447678  803.739136  750.023376    0.658006      5     bus
4    0.000000  550.596008   76.681190  878.669922    0.450596      0  person
```

This opens the doors to some JSON export options using the pandas .to_json() methods. The results.pandas() method exports all torch tensor results to dataframes at once (i.e. for batch-size 10 there will be 10 torch.tensors / pd.DataFrames), but note that each image would need to run it's own .to_json(), i.e. for the first image:

```python
results.pandas().xyxy[0].to_json()
'{"xmin":{"0":57.0689697266,"1":667.6612548828,"2":222.8783874512,"3":4.2053861618,"4":0.0},"ymin":{"0":391.7705993652,"1":399.3035888672,"2":414.774230957,"3":234.4476776123,"4":550.5960083008},"xmax":{"0":241.3835449219,"1":810.0,"2":343.804473877,"3":803.7391357422,"4":76.6811904907},"ymax":{"0":905.7978515625,"1":881.3966674805,"2":857.8250732422,"3":750.0233764648,"4":878.669921875},"confidence":{"0":0.8689641356,"1":0.8518877029,"2":0.8383761048,"3":0.6580058336,"4":0.4505961835},"class":{"0":0.0,"1":0.0,"2":0.0,"3":5.0,"4":0.0},"name":{"0":"person","1":"person","2":"person","3":"bus","4":"person"}}'
    
# or probably better:
results.pandas().xyxy[0].to_json(orient='records')
'[{"xmin":57.0689697266,"ymin":391.7705993652,"xmax":241.3835449219,"ymax":905.7978515625,"confidence":0.8689641356,"class":0.0,"name":"person"},{"xmin":667.6612548828,"ymin":399.3035888672,"xmax":810.0,"ymax":881.3966674805,"confidence":0.8518877029,"class":0.0,"name":"person"},{"xmin":222.8783874512,"ymin":414.774230957,"xmax":343.804473877,"ymax":857.8250732422,"confidence":0.8383761048,"class":0.0,"name":"person"},{"xmin":4.2053861618,"ymin":234.4476776123,"xmax":803.7391357422,"ymax":750.0233764648,"confidence":0.6580058336,"class":5.0,"name":"bus"},{"xmin":0.0,"ymin":550.5960083008,"xmax":76.6811904907,"ymax":878.669921875,"confidence":0.4505961835,"class":0.0,"name":"person"}]'
```

What do you think?

**Comment by robmarkcole** [comment]
Hi @glenn-jocher that looks great, and the nice thing about going the pandas route is that a user can manipulate the table to return json in any format they want before export as json

**Comment by glenn-jocher** [comment]
@robmarkcole PR #2725 is merged. You can now access any of the results as pandas DataFrames with results.pandas(). The four outputs that are converted are results.xyxy, .xyxyn, .xywh, .xywhn. The DataFrame columns are updated depending on the output type xyxy or xywh.

I think this is a great change, we can also now use the pandas table to visualize the results better in the YOLOv5 tutorials such as https://pytorch.org/hub/ultralytics_yolov5.

I also updated the class column to integer format to clean it up a bit.

NOTE: The pandas json exports seem to carry far more significant figures than necessary. These might be slimmed down a bit to reduce bandwidth on API traffic, perhaps to `{:,.5f}.format` for floats.

**Comment by glenn-jocher** [comment]
Removing TODO now as this feature is implemented in #2725.

**Comment by ozett** [comment]
hard to find a simple docu on this site, of
how to get the names from predictions as a n00b-python coder.
but glad this was here. 
thanks for this ```print (results.pandas().xyxy[0].name.to_json())``` 
💐
